### PR TITLE
[Merged by Bors] - Document alignment requirements of `Ptr`, `PtrMut` and `OwningPtr`

### PR DESCRIPTION
**Pull request by BoxyUwU**
# Objective

The types in the `bevy_ptr` accidentally did not document anything relating to alignment. This is unsound as many methods rely on the pointer being correctly aligned. 

## Solution

This PR introduces new safety invariants on the `$ptr::new`, `$ptr::byte_offset` and `$ptr::byte_add` methods requiring them to keep the pointer aligned. This is consistent with the documentation of these pointer types which document them as being "type erased borrows".

As it was pointed out (by @JoJoJet in #7117) that working with unaligned pointers can be useful (for example our commands abstraction which does not try to align anything properly, see #7039) this PR also introduces a default type parameter to all the pointer types that specifies whether it has alignment requirements or not. I could not find any code in `bevy_ecs` that would need unaligned pointers right now so this is going unused.

---

## Changelog

- Correctly document alignment requirements on `bevy_ptr` types.
- Support variants of `bevy_ptr` types that do not require being correctly aligned for the pointee type.

## Migration Guide

- Safety invariants on `bevy_ptr` types' `new` `byte_add` and `byte_offset` methods have been changed. All callers should re-audit for soundness.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
Nice use of the type system to uphold this.

File: crates/bevy_ptr/src/lib.rs [hunk 1]
```diff
@@ -7,17 +7,37 @@ use core::{
...
+/// Used as a type argument to [`Ptr`], [`PtrMut`] and [`OwningPtr`] to specify that the pointer is aligned.
+pub struct Aligned;
+#[derive(Copy, Clone)]
+/// Used as a type argument to [`Ptr`], [`PtrMut`] and [`OwningPtr`] to specify that the pointer is not aligned.
+pub struct Unaligned;
+
+/// Trait that is only implemented for [`Aligned`] and [`Unaligned`] to work around the lack of ability
+/// to have const generics of an enum.
+pub trait IsAligned: sealed::Sealed {}
+impl IsAligned for Aligned {}
```
Comment (james7132) [inline crates/bevy_ptr/src/lib.rs:20]: Actually wonder if we could use const generics to enforce debug checks on `byte_add` and `byte_offset`, but I'll leave that for a separate PR.

**Review by james7132 — APPROVED** [review]

File: crates/bevy_ptr/src/lib.rs (hunk 1 above)
Comment (BoxyUwU) [reply crates/bevy_ptr/src/lib.rs:20]: you wouldnt need const generics, it ought to be possible to look at the `TypeId` of `A` to see if its `Aligned` or not

**Comment by BoxyUwU** [comment]
Interesting note: the original PR that added the `Ptr`/`PtrMut`/`OwningPtr` types #3001 originally did have documentation stating it was guaranteed to be aligned (although was missing safety invariants for it) but got removed when more docs were added :sweat_smile:  https://github.com/bevyengine/bevy/pull/3001/commits/b76d47ae6d91363b8bd2c924a074c66c87ec60ac#diff-d06ecda1c4e3f50921be6f83a8be514f511ce6129d9f872bc4015d3bdf75c0c5L3

**Pull request closed without merging**